### ISSUE TITLE
[GPU] [performance] LightGBM Running Significantly Slower Than XGBoost on GPU for Classification Task

### ISSUE THREAD

**Issue by Soontosh** [microsoft/LightGBM#6531]
### Context
I am applying `LightGBMClassifier` for the classification of DNA Strands from Electrical Conductance Traces. Input data is represented as a Histogram with 600 Bins. There are three classes and 5250 samples in the training set.

### System Information
- Operating System: WSL-2 Ubuntu
- LightGBM Version: 4.4.0.99 (built with `DUSE_CUDA=1`)
- CUDA Version 12.5
- 32 GB RAM, Notebook RTX Nvidia 4060

### The Issue
I have tested both XGBoost and LightGBM for this task, they achieve similar performance in regard to accuracy.  When testing their speed, I found XGBoost to generate predictions ~3x faster than LightGBM, given testing samples. I ran the test many times, and have consistently gotten this result. Below are the console logs resulting from tests:
```
XGBoostClassifier
Time taken for 112500 samples:  0.151678955  seconds.
Time taken for 1 sample:  1.3482573777777778e-06  seconds.

LightGBMClassifier
Time taken for 112500 samples:  0.327497704  seconds.
Time taken for 1 sample:  2.9110907022222224e-06  seconds.
```

This contradicts everything I have read in the literature about these two architectures. I am wondering if I am doing something wrong with how I set up LightGBM, or if there are more optimizations I can be applying to improve its performance.
```python
""" LightGBM Setup """
# Compiling Model
lgbm = lgb.LGBMClassifier(n_estimators=200, num_leaves=3, max_depth=2, verbose=-1)
lgbm.set_params(device='cuda')

# Predicting Labels (test_data is an ArrayLike with all the testing samples)
return self.model.predict(test_data)

""" XGBoost Setup """
# Compiling Model
params = dict()
params["device"] = "cuda"
xgb = XGBClassifier(n_estimators=200, max_depth=2, verbosity = 0, params=params, thread_count=-1)

# Predicting Labels (test_data is an ArrayLike with all the testing samples)
return self.model.predict(test_data)
```

I would like to note that I have already tried tweaking the  `num_leaves` and `max_depth` parameters of my LightGBM model, which started as `32` and `-1` respectively. I have seen a very small boost in speed after doing this, but it was not enough to make much of a difference.

Thank you so much for your help!

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

To be clear, are you saying that **training** is that much slower? Or **prediction**? I noticed your sample code does not call `.fit()`, and I wonder if by "reach a classification" you mean "generate predictions".

**Comment by Soontosh** [comment]
> Thanks for using LightGBM.
> 
> To be clear, are you saying that **training** is that much slower? Or **prediction**? I noticed your sample code does not call `.fit()`, and I wonder if by "reach a classification" you mean "generate predictions".

Hello, thank you for your reply! To clarify, I was referring to the time it took to "generate predictions" given a sample, not how long the model took to train. I'll update my question with the correct wording.

**Comment by jameslamb** [comment]
Thanks for that clarification.

I believe that LightGBM doesn't currently have CUDA-accelerated prediction support, so the time it takes to generate predictions will not be affected by specifying `device="cuda"`.

@shiyu1994 am I right about that?

You could see some other examples for how to use GPU acceleration for generating predictions from a LightGBM model: https://github.com/microsoft/LightGBM/issues/5854#issuecomment-2138659914

**Comment by Soontosh** [comment]
Oh I see, that is unfortunate. This seems to be true - I just disabled the CUDA device and saw no slowdown in performance. Thank you for your help, I'll take a look at the libraries you mentioned.

**Comment by Soontosh** [comment]
@jameslamb Found this tool by Microsoft specifically for fixing this: https://opensource.microsoft.com/blog/2020/09/29/accelerate-machine-learning-models-gpu-onnx-runtime-hummingbird/. I think this should be mentioned somewhere in the docs, maybe in the GPU setup guide?

**Comment by jameslamb** [comment]
> I think this should be mentioned somewhere in the docs

Oh interesting, thanks for the link!

It is actually mentioned in LightGBMs's main README:

https://github.com/microsoft/LightGBM/blob/3d386be29bb9e8032e9acaec7c831133928fbeba/README.md#L82

But I understand that it's not obvious that "model compiler into tensor computations" implies "use GPUs to run inference faster". I will think about where best to put this in the documentation. There are enough complex features around scoring here (prediction early stopping, CPU parallelism, GPU acceleration, fast single-row prediction vs. higher-throughput batch prediction) that it might be worthwhile to have an entire page devoted just to those topics.
